File: packages/autonomo-python/autonomo/instance.py
"""
Instance - Unique identity for this app instance

Each window/tab/simulator gets a unique instance ID that persists
for the lifetime of that instance. This allows the MCP server to
distinguish between multiple instances of the same app.
"""

from dataclasses import dataclass, field
from typing import Optional, Dict, Any, Literal
import time
import uuid


Platform = Literal["web", "mobile", "desktop"]


@dataclass
class InstanceConfig:
    """Configuration for initializing an app instance"""
    name: str
    platform: Platform
    instance_id: Optional[str] = None
    version: Optional[str] = None
    meta: Optional[Dict[str, Any]] = None


@dataclass
class InstanceInfo:
    """Information about this app instance"""
    instance_id: str
    name: str
    bridge_id: str
    platform: Platform
    version: Optional[str] = None
    created_at: int = field(default_factory=lambda: int(time.time() * 1000))
    meta: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        result = {
            "instanceId": self.instance_id,
            "name": self.name,
            "bridgeId": self.bridge_id,
            "platform": self.platform,
            "createdAt": self.created_at,
        }
        if self.version is not None:
            result["version"] = self.version
        if self.meta is not None:
            result["meta"] = self.meta
        return result


class InstanceManager:
    """Singleton instance manager"""
    
    _instance: Optional["InstanceManager"] = None
    _current: Optional[InstanceInfo] = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance
    
    @staticmethod
    def _generate_instance_id() -> str:
        """Generate a short unique ID"""
        return uuid.uuid4().hex[:8]
    
    def init_instance(self, config: InstanceConfig) -> InstanceInfo:
        """
        Initialize this app instance.
        
        Call once at app startup. Each process gets a unique instance ID.
        
        Example:
            instance_manager.init_instance(InstanceConfig(
                name="my-app",
                platform="web"
            ))
        """
        instance_id = config.instance_id or self._generate_instance_id()
        self._current = InstanceInfo(
            instance_id=instance_id,
            name=config.name,
            bridge_id=f"{config.name}-{instance_id}",
            platform=config.platform,
            version=config.version,
            meta=config.meta,
        )
        print(f"[Autonomo] Instance initialized: {self._current.bridge_id}")
        return self._current
    
    def get_instance(self) -> Optional[InstanceInfo]:
        """Get the current instance info"""
        return self._current
    
    def require_instance(self) -> InstanceInfo:
        """Get the current instance info or raise"""
        if self._current is None:
            raise RuntimeError("Autonomo instance not initialized. Call init_instance() first.")
        return self._current
    
    def get_bridge_id(self) -> Optional[str]:
        """Get just the bridge ID"""
        return self._current.bridge_id if self._current else None
    
    def reset_instance(self) -> None:
        """Reset the instance (mainly for testing)"""
        self._current = None


# Global instance manager singleton
instance_manager = InstanceManager()


# Convenience functions
def init_instance(config: InstanceConfig) -> InstanceInfo:
    """Initialize this app instance"""
    return instance_manager.init_instance(config)


def get_instance() -> Optional[InstanceInfo]:
    """Get the current instance info"""
    return instance_manager.get_instance()


def require_instance() -> InstanceInfo:
    """Get the current instance info or raise"""
    return instance_manager.require_instance()


def get_bridge_id() -> Optional[str]:
    """Get just the bridge ID"""
    return instance_manager.get_bridge_id()


def reset_instance() -> None:
    """Reset the instance (mainly for testing)"""
    instance_manager.reset_instance()

File: packages/autonomo_flutter/lib/src/instance.dart
/// Instance - Unique identity for this app instance
///
/// Each window/tab/simulator gets a unique instance ID that persists
/// for the lifetime of that instance. This allows the MCP server to
/// distinguish between multiple instances of the same app.

import 'dart:math';

/// Platform type for the instance
enum Platform { web, mobile, desktop }

/// Configuration for initializing an app instance
class InstanceConfig {
  final String name;
  final Platform platform;
  final String? instanceId;
  final String? version;
  final Map<String, dynamic>? meta;

  const InstanceConfig({
    required this.name,
    required this.platform,
    this.instanceId,
    this.version,
    this.meta,
  });
}

/// Information about this app instance
class InstanceInfo {
  final String instanceId;
  final String name;
  final String bridgeId;
  final Platform platform;
  final String? version;
  final int createdAt;
  final Map<String, dynamic>? meta;

  const InstanceInfo({
    required this.instanceId,
    required this.name,
    required this.bridgeId,
    required this.platform,
    this.version,
    required this.createdAt,
    this.meta,
  });

  Map<String, dynamic> toJson() => {
        'instanceId': instanceId,
        'name': name,
        'bridgeId': bridgeId,
        'platform': platform.name,
        'createdAt': createdAt,
        if (version != null) 'version': version,
        if (meta != null) 'meta': meta,
      };
}

/// Singleton instance manager
class InstanceManager {
  static final InstanceManager _instance = InstanceManager._internal();
  factory InstanceManager() => _instance;
  InstanceManager._internal();

  InstanceInfo? _currentInstance;

  /// Generate a short unique ID
  String _generateInstanceId() {
    const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
    final random = Random.secure();
    return List.generate(8, (_) => chars[random.nextInt(chars.length)]).join();
  }

  /// Initialize this app instance
  ///
  /// Call once at app startup. Each window/process gets a unique instance ID.
  ///
  /// ```dart
  /// void main() {
  ///   instanceManager.initInstance(InstanceConfig(
  ///     name: 'my-app',
  ///     platform: Platform.mobile,
  ///   ));
  ///   runApp(MyApp());
  /// }
  /// ```
  InstanceInfo initInstance(InstanceConfig config) {
    final instanceId = config.instanceId ?? _generateInstanceId();
    _currentInstance = InstanceInfo(
      instanceId: instanceId,
      name: config.name,
      bridgeId: '${config.name}-$instanceId',
      platform: config.platform,
      version: config.version,
      createdAt: DateTime.now().millisecondsSinceEpoch,
      meta: config.meta,
    );
    print('[Autonomo] Instance initialized: ${_currentInstance!.bridgeId}');
    return _currentInstance!;
  }

  /// Get the current instance info
  InstanceInfo? getInstance() => _currentInstance;

  /// Get the current instance info or throw
  InstanceInfo requireInstance() {
    if (_currentInstance == null) {
      throw StateError('Autonomo instance not initialized. Call initInstance() first.');
    }
    return _currentInstance!;
  }

  /// Get just the bridge ID
  String? getBridgeId() => _currentInstance?.bridgeId;

  /// Reset the instance (mainly for testing)
  void resetInstance() {
    _currentInstance = null;
  }
}

/// Global instance manager singleton
final instanceManager = InstanceManager();

/// Convenience function to initialize instance
InstanceInfo initInstance(InstanceConfig config) => instanceManager.initInstance(config);

/// Convenience function to get instance
InstanceInfo? getInstance() => instanceManager.getInstance();

/// Convenience function to require instance
InstanceInfo requireInstance() => instanceManager.requireInstance();

/// Convenience function to get bridge ID
String? getBridgeId() => instanceManager.getBridgeId();

/// Convenience function to reset instance
void resetInstance() => instanceManager.resetInstance();
